core/devdoc/on-out-process-gateway-modules.md

```diff
@@ -0,0 +1,484 @@
+On out process gateway modules
+==============================
+
+Introduction
+------------
+
+All modules in the gateway are hosted within the gateway process today. While
+this makes for great performance it is also at the cost of execution isolation
+between modules, i.e., one misbehaving module can potentially affect the
+behavior of other modules or the gateway itself. This document seeks to propose
+a design for implementing gateway modules that are hosted in external processes.
+These processes will run intially on the same device that the gateway is 
+running on, with the option of extending this behavior to remote devices.
+
+The outprocess module is applicable for a customer that already has device 
+telemetry and control in another process, and wants to send that telemetry data 
+to the gateway with a minimum of new code.
+
+Design Goals
+------------
+
+One design goal is the notion that developers who are implementing modules
+should be able to take a working module as is and host the module in a process
+outside the gateway process. All the complexities of activation and
+inter-process communication should be transparently handled by the gateway
+framework.
+
+Another design goal is to enable the implementation of module host 
+processes in popular programming languages and technology stacks. This is 
+will be achieved by specifying the serialized gateway and control 
+messages and enumerating the expectations of a module host implementation.
+
+Design Non-goals
+----------------
+
+It is possible that an outprocess module, due to implementation issues (aka 
+bugs), becomes unstable and precipitates a process hang or termination. The 
+scope of this document does not cover these recovery scenarios. It is assumed that mechanisms external to the gateway will be used for this purpose at this 
+time.
+
+Design Overview
+---------------
+
+![Design Overview](./media/outprocess-gateway-modules.png)
+
+The Gateway will still consist of the gateway, broker and modules, seen above as the "Gateway Process."  To implement the out of process modules, the gateway process will now have a proxy module which will forward control and data to a module host process.
+
+As an example of control flow from the Gateway to the Module host, this is an elided sequence diagram for outprocess module creation:
+
+![](./media/gateway_oop_create.png)
+
+NOTE: The call to the out of process module will be a synchronous call.
```

Comment:
Right now, it's waiting forever since we don't know when the remote process is going to show up or choose to respond.  I think this need to be used & tweaked before we know what the best choices are.